java/driver/jni/CMakeLists.txt

```diff
@@ -0,0 +1,44 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+find_package(AdbcDriverManager)
+
+add_custom_command(OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/target/headers/org_apache_arrow_adbc_driver_jni_impl_NativeAdbc.h
+                   COMMENT "Generate JNI headers"
+                   # Force Maven to actually re-run the command
+                   COMMAND rm -rf ${CMAKE_CURRENT_SOURCE_DIR}/target/headers
+                           ${CMAKE_CURRENT_SOURCE_DIR}/target/maven-status
+                   COMMAND mvn --file ${CMAKE_CURRENT_SOURCE_DIR}/../.. -Pjni,javah
+                           compile --also-make --projects :adbc-driver-jni
+                   DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/src/main/java/org/apache/arrow/adbc/driver/jni/impl/NativeAdbc.java
+)
+
+add_library(adbc_driver_jni SHARED
+            src/main/cpp/jni_wrapper.cc
+            ${CMAKE_CURRENT_SOURCE_DIR}/target/headers/org_apache_arrow_adbc_driver_jni_impl_NativeAdbc.h
+)
```

Comment:
CMake projects `add_jar()` https://cmake.org/cmake/help/latest/module/UseJava.html#creating-and-installing-jars .
Does it help this?

FYI: apache/arrow-java uses `add_jar()` when I prepared it: https://github.com/apache/arrow-java/blob/d304da5f872615346557e947ca5b04e9e23e8b87/dataset/CMakeLists.txt#L24-L32

Reply:
That's what I originally did! :joy: 

But it has some problems. CMake can't resolve the Maven dependency tree. So the files we pass to `add_jar` have to be able to compile without any dependencies, including any dependencies on other ADBC classes. That meant duplicating code or writing slightly unnatural code in Java. So @laurentgo here suggested using CMake to invoke Maven to generate the headers instead.

Reply:
Oh, sorry!